File: src/test/java/io/binac/leetcode/RecoverBinarySearchTreeTest.java
package io.binac.leetcode;

import io.binac.leetcode.util.TreeNode;
import io.binac.leetcode.util.Trees;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class RecoverBinarySearchTreeTest {
    private final RecoverBinarySearchTree.Solution1 solution1 = new RecoverBinarySearchTree.Solution1();

    @Test
    void test1() {
        TreeNode root1 = Trees.asTree(1, 3, null, null, 2);
        solution1.recoverTree(root1);
        TreeNode expected1 = Trees.asTree(3, 1, null, null, 2);
        assertTrue(Trees.equals(expected1, root1));

        TreeNode root2 = Trees.asTree(3, 1, 4, null, null, 2);
        solution1.recoverTree(root2);
        TreeNode expected2 = Trees.asTree(2, 1, 4, null, null, 3);
        assertTrue(Trees.equals(expected2, root2));
    }
}

File: cpp/tests/RecoverBinarySearchTreeTest.cpp
#include "RecoverBinarySearchTree.h"
#include "gtest/gtest.h"

using namespace lcpp;

TEST(RecoverBinarySearchTree, Solution99_1) {
  auto S99_1 = Solution99_1();

  auto Root1 = newTree({1, 3, NTNode, NTNode, 2});
  auto Expected1 = newTree({3, 1, NTNode, NTNode, 2});
  S99_1.recoverTree(Root1);
  EXPECT_EQ(*Expected1, *Root1);
  deleteTree(Root1);
  deleteTree(Expected1);

  auto Root2 = newTree({3, 1, 4, NTNode, NTNode, 2});
  auto Expected2 = newTree({2, 1, 4, NTNode, NTNode, 3});
  S99_1.recoverTree(Root2);
  EXPECT_EQ(*Expected2, *Root2);
  deleteTree(Root2);
  deleteTree(Expected2);
}

TEST(RecoverBinarySearchTree, Solution99_2) {
  auto S99_2 = Solution99_2();

  auto Root1 = newTree({1, 3, NTNode, NTNode, 2});
  auto Expected1 = newTree({3, 1, NTNode, NTNode, 2});
  S99_2.recoverTree(Root1);
  EXPECT_EQ(*Expected1, *Root1);
  deleteTree(Root1);
  deleteTree(Expected1);

  auto Root2 = newTree({3, 1, 4, NTNode, NTNode, 2});
  auto Expected2 = newTree({2, 1, 4, NTNode, NTNode, 3});
  S99_2.recoverTree(Root2);
  EXPECT_EQ(*Expected2, *Root2);
  deleteTree(Root2);
  deleteTree(Expected2);
}